src/dotnet-install.sh

```diff
@@ -541,43 +605,87 @@ construct_download_link() {
+    # if 'productVersion.txt' is missing but download link is already available, product version will be taken from download link
     eval $invocation
 
     local azure_feed="$1"
     local specific_version="${2//[$'\t\r\n']}"
+    local package_download_link=""
+    if [ $# -gt 2  ]; then
+        local package_download_link="$3"
+    fi
     local specific_product_version=$specific_version
 
     local download_link=null
-    if [[ "$runtime" == "dotnet" ]]; then
-        download_link="$azure_feed/Runtime/$specific_version/productVersion.txt${feed_credential}"
-    elif [[ "$runtime" == "aspnetcore" ]]; then
-        download_link="$azure_feed/aspnetcore/Runtime/$specific_version/productVersion.txt${feed_credential}"
-    elif [ -z "$runtime" ]; then
-        download_link="$azure_feed/Sdk/$specific_version/productVersion.txt${feed_credential}"
+
+    if [ -z "$package_download_link" ]; then
+        if [[ "$runtime" == "dotnet" ]]; then
+            download_link="$azure_feed/Runtime/$specific_version/productVersion.txt${feed_credential}"
+        elif [[ "$runtime" == "aspnetcore" ]]; then
+            download_link="$azure_feed/aspnetcore/Runtime/$specific_version/productVersion.txt${feed_credential}"
+        elif [ -z "$runtime" ]; then
+            download_link="$azure_feed/Sdk/$specific_version/productVersion.txt${feed_credential}"
+        else
+            return 1
+        fi
     else
-        return 1
+        download_link="${package_download_link%/*}/productVersion.txt${feed_credential}" 
     fi
 
     if machine_has "curl"
     then
         specific_product_version=$(curl -s --fail "$download_link")
         if [ $? -ne 0 ]
         then
-            specific_product_version=$specific_version
+            specific_product_version="$(get_product_specific_version_from_download_link "$package_download_link" "$specific_version")"
         fi
     elif machine_has "wget"
     then
         specific_product_version=$(wget -qO- "$download_link")
         if [ $? -ne 0 ]
         then
-            specific_product_version=$specific_version
+            specific_product_version="$(get_product_specific_version_from_download_link "$package_download_link" "$specific_version")"
         fi
     fi
     specific_product_version="${specific_product_version//[$'\t\r\n']}"
+    echo "$specific_product_version"
+    return 0
+}
 
+# args:
+# download link - $1
+# specific version - $2
+get_product_specific_version_from_download_link()
+{
+    eval $invocation
+
+    local download_link="$1"
+    local specific_version="$2"
+    local specific_product_version="" 
+
+    if [ ! -z "$download_link" ]; then
+        echo "$specific_version"
+        return 0
+    fi
+
+    #get filename
+    filename="${download_link##*/}"
+
+    #product specific version follows the product name
+    #for filename 'dotnet-sdk-3.1.404-linux-x64.tar.gz': the product version is 3.1.400
```

Comment:
Which line in this code block is converting 3.1.404 into 3.1.400?

Reply:
that's a typo :)